supertokens_python/recipe/session/jwt.py

```diff
@@ -46,39 +47,72 @@
 class ParsedJWTInfo:
     def __init__(
         self,
+        version: int,
         raw_token_string: str,
         raw_payload: str,
         header: str,
         payload: Dict[str, Any],
         signature: str,
+        kid: Optional[str],
     ) -> None:
+        self.version = version
         self.raw_token_string = raw_token_string
         self.raw_payload = raw_payload
         self.header = header
         self.payload = payload
         self.signature = signature
+        self.kid = kid
 
 
 def parse_jwt_without_signature_verification(jwt: str) -> ParsedJWTInfo:
     splitted_input = jwt.split(".")
     if len(splitted_input) != 3:
         raise Exception("invalid jwt")
 
+    # V1 and V2 are functionally identical, plus all legacy tokens should be V2 now.
+    # So we can assume these defaults:
+    version = 2
+    kid = None
+    # V2 or older tokens didn't save the key id
     header, payload, signature = splitted_input
+    # checking the header
     if header not in _allowed_headers:
-        raise Exception("jwt header mismatch")
+        parsed_header = json.loads(base64.b64decode(header.encode()))
+        header_version = parsed_header.get("version")
+
+        # We have to ensure version is a string, otherwise Number.parseInt can have unexpected results
+        if not isinstance(header_version, str):
+            raise Exception("JWT header mismatch")
+
+        try:
+            version = int(header_version)
+        except ValueError:
+            version = None
+
+        kid = parsed_header.get("kid")
+        # Number.isInteger returns false for Number.NaN (if it fails to parse the version)
+        if (
+            parsed_header["typ"] != "JWT"
+            or not isinstance(version, int)
+            or version < 3
+            or kid is None
+        ):
+            raise Exception("JWT header mismatch")
 
     return ParsedJWTInfo(
+        version=version,
         raw_token_string=jwt,
         raw_payload=payload,
         header=header,
         # Ideally we would only parse this after the signature verification is done
         # We do this at the start, since we want to check if a token can be a supertokens access token or not.
         payload=loads(utf_base64decode(payload)),
         signature=signature,
+        kid=kid,  # TODO: How can ever be None since we are checking for it in the above code?
```

Comment:
TODO comment.
Btw, this is `None` if we have a V2 token (in which case `header not in _allowed_headers` is false)